File: backend/app/schemas/project.py
"""Project related Pydantic schemas."""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class ProjectCreateDTO(BaseModel):
    """DTO for creating a project."""

    project_id: str | None = Field(None, description="Optional project ID")
    project_name: str = Field(..., max_length=64)
    project_description: str | None = Field(None, max_length=256)
    template_level: str = Field(...)


class ProjectUpdateDTO(BaseModel):
    """DTO for updating a project."""

    project_name: str | None = Field(None, max_length=64)
    project_description: str | None = Field(None, max_length=256)


class ProjectExecutionStrategyUpdateDTO(BaseModel):
    """DTO for updating a project's execution strategy."""

    execution_strategy: str = Field(
        ..., description="执行策略: full_auto / semi_auto / full_manual"
    )
    reason: str | None = Field(None, description="变更原因")


class ProjectResponseDTO(BaseModel):
    """DTO for project response."""

    model_config = ConfigDict(from_attributes=True)

    project_id: str
    project_name: str
    project_description: str | None
    project_status: str
    application_id: str
    template_level: str
    progress_percent: int
    current_stage: str | None
    current_stage_id: str | None
    risk_level: str
    last_activity_at: str | None = None
    last_activity_type: str | None
    size_estimate_id: str | None
    execution_strategy: str = "semi_auto"
    merge_policy_json: dict[str, Any] | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None


class RiskAlertDTO(BaseModel):
    """DTO for a risk alert."""

    alert_type: str
    severity: str
    message: str
    project_id: str | None = None
    stage_id: str | None = None


class TimeboxEntryDTO(BaseModel):
    """DTO for a timebox entry."""

    stage_id: str
    stage_name: str
    planned_days: float
    elapsed_days: float
    remaining_days: float
    deviation_percent: float
    alert_level: str | None


class StageProgressDTO(BaseModel):
    """DTO for project stage progress."""

    stage_id: str
    stage_name: str
    order_index: int
    status: str
    runtime_status: str = "not_started"
    execution_status: str
    progress_percent: int = Field(default=0, ge=0, le=100)
    planned_days: float | None = None
    elapsed_days: float | None = None
    skippable: bool = False


class ArtifactSummaryDTO(BaseModel):
    """DTO for artifact summary in project detail."""

    artifact_id: str
    file_name: str
    file_type: str
    stage_id: str | None
    created_at: str | None


class OperationLogItemDTO(BaseModel):
    """DTO for operation log item in project detail."""

    log_id: str
    action: str
    operator_id: str | None
    target_type: str | None
    detail: str | None
    created_at: str | None


class SizeEstimateResultDTO(BaseModel):
    """DTO for size estimate result bound to project."""

    estimate_id: str | None
    module_count: int | None
    interface_count: int | None
    page_count: int | None
    tech_complexity: str | None
    risk_level: str | None
    optimistic_score: int | None
    expected_score: int | None
    conservative_score: int | None
    complexity_level: str | None


class ProjectOverviewDTO(BaseModel):
    """DTO for project overview in detail drawer."""

    project: ProjectResponseDTO
    size_estimate: SizeEstimateResultDTO | None
    stages: list[StageProgressDTO]
    artifacts: list[ArtifactSummaryDTO]
    operation_logs: list[OperationLogItemDTO]


class BindSizeEstimateDTO(BaseModel):
    """DTO for binding size estimate to project."""

    estimate_id: str | None = Field(None, description="规模评估ID，null表示解绑")


class StageStartResponseDTO(BaseModel):
    """DTO for starting a project stage pipeline."""

    project_id: str
    current_stage_id: str
    status: str


class ProjectExecutionStrategyResponseDTO(BaseModel):
    """DTO for project execution strategy update response."""

    project_id: str
    execution_strategy: str
    updated_stage_ids: list[str] = Field(default_factory=list)


class StageExecuteResponseDTO(BaseModel):
    """DTO for executing a project stage."""

    project_stage_id: str
    execution_ids: list[str]
    status: str
    next_stage_id: str | None = None


class StageAdvanceResponseDTO(BaseModel):
    """DTO for advancing a project stage."""

    project_stage_id: str
    status: str
    next_stage_id: str | None = None


class StageGateDecisionDTO(BaseModel):
    """DTO for Gate decision."""

    decision: str = Field(..., description="pass 或 reject")
    reason: str | None = Field(None, description="驳回时必填")


class StageGateDecisionResponseDTO(BaseModel):
    """DTO for Gate decision response."""

    project_stage_id: str
    status: str
    next_stage_id: str | None = None


class StageProgressResponseDTO(BaseModel):
    """DTO for aggregated project stage progress."""

    project_id: str
    execution_strategy: str
    current_stage_id: str | None
    progress_percent: int
    stages: list[dict[str, Any]]


class StageRollbackRequestDTO(BaseModel):
    """DTO for stage rollback request."""

    target_stage_id: str = Field(..., description="回滚目标阶段 ID")
    reason: str | None = Field(None, description="回滚原因")


class StageRollbackResponseDTO(BaseModel):
    """DTO for stage rollback response."""

    project_id: str
    target_stage_id: str
    reset_stage_ids: list[str] = Field(default_factory=list)
    stale_artifact_ids: list[str] = Field(default_factory=list)

File: frontend/src/services/project.ts
import api from './api'

export interface PageResponse<T> {
  data: T[]
  total_count: number
  page: number
  page_size: number
  total_pages: number
  has_next: boolean
  has_previous: boolean
}

export interface ApplicationItem {
  application_id: string
  application_name: string
  description: string | null
  local_path: string
  workspace_id: string
  path_accessible: boolean
}

export interface Project {
  project_id: string
  project_name: string
  project_description: string | null
  project_status: 'Draft' | 'Active' | 'Archived' | 'Cancelled'
  application_id: string
  template_level: string
  progress_percent: number
  current_stage: string | null
  risk_level: 'None' | 'Low' | 'Medium' | 'High'
  last_activity_at: string | null
  last_activity_type: string | null
  size_estimate_id: string | null
  created_at: string
  updated_at: string
}

export interface RiskAlert {
  alert_type: string
  severity: 'Low' | 'Medium' | 'High'
  message: string
  project_id: string | null
  stage_id: string | null
}

export interface TimeboxEntry {
  stage_id: string
  stage_name: string
  planned_days: number
  elapsed_days: number
  remaining_days: number
  deviation_percent: number
  alert_level: string | null
}

export interface ProjectCreatePayload {
  project_id?: string
  project_name: string
  project_description?: string | null
  template_level: string
}

export interface ProjectUpdatePayload {
  project_name?: string
  project_description?: string | null
}

export interface StageProgress {
  stage_id: string
  stage_name: string
  order_index: number
  status: string
  execution_status: string
  progress_percent: number
  planned_days: number | null
  elapsed_days: number | null
  skippable: boolean
}

export interface ArtifactSummary {
  artifact_id: string
  file_name: string
  file_type: string
  stage_id: string | null
  created_at: string | null
}

export interface OperationLogItem {
  log_id: string
  action: string
  operator_id: string | null
  target_type: string | null
  detail: string | null
  created_at: string | null
}

export interface SizeEstimateResult {
  estimate_id: string | null
  module_count: number | null
  interface_count: number | null
  page_count: number | null
  tech_complexity: string | null
  risk_level: string | null
  optimistic_score: number | null
  expected_score: number | null
  conservative_score: number | null
  complexity_level: string | null
}

export interface ProjectOverview {
  project: Project
  size_estimate: SizeEstimateResult | null
  stages: StageProgress[]
  artifacts: ArtifactSummary[]
  operation_logs: OperationLogItem[]
}

export interface ComplexityAssessPayload {
  module_count: number
  interface_complexity: number
  page_count: number
  entity_count: number
  integration_count: number
}

export interface ComplexityAssessResult {
  optimistic_score: number
  expected_score: number
  conservative_score: number
  complexity_level: string
  radar_values: Record<string, number> | null
}

export interface SizeEstimatePayload {
  module_count: number
  interface_count: number
  page_count: number
  tech_complexity: string
  risk_level: string
}

export interface SizeEstimateResponse {
  estimate_id: string
  project_id: string
  module_count: number
  interface_count: number
  page_count: number
  tech_complexity: string
  risk_level: string
  optimistic_score: number | null
  expected_score: number | null
  conservative_score: number | null
  complexity_level: string | null
  created_at: string | null
}

export async function fetchProjects(appId: string): Promise<Project[]> {
  const res = await api.get(`/v1/applications/${appId}/projects`)
  return (res.data as { data?: Project[] })?.data ?? []
}

export async function createProject(
  appId: string,
  payload: ProjectCreatePayload,
): Promise<Project> {
  const res = await api.post<Project>(`/v1/applications/${appId}/projects`, payload)
  return res.data
}

export async function getProject(projectId: string): Promise<Project> {
  const res = await api.get<Project>(`/v1/projects/${projectId}`)
  return res.data
}

export async function updateProject(
  projectId: string,
  payload: ProjectUpdatePayload,
): Promise<Project> {
  const res = await api.patch<Project>(`/v1/projects/${projectId}`, payload)
  return res.data
}

export async function archiveProject(projectId: string): Promise<void> {
  await api.post(`/v1/projects/${projectId}/archive`)
}

export async function activateProject(projectId: string): Promise<Project> {
  const res = await api.post<Project>(`/v1/projects/${projectId}/activate`)
  return res.data
}

export async function cancelProject(projectId: string): Promise<Project> {
  const res = await api.post<Project>(`/v1/projects/${projectId}/cancel`)
  return res.data
}

export async function fetchRiskAlerts(projectId: string): Promise<RiskAlert[]> {
  const res = await api.get<RiskAlert[]>(`/v1/projects/${projectId}/risk-alerts`)
  return res.data
}

export async function fetchTimebox(projectId: string): Promise<TimeboxEntry[]> {
  const res = await api.get<TimeboxEntry[]>(`/v1/projects/${projectId}/timebox`)
  return res.data
}

export async function fetchProjectOverview(projectId: string): Promise<ProjectOverview> {
  const res = await api.get<ProjectOverview>(`/v1/projects/${projectId}/overview`)
  return res.data
}

export async function fetchProjectStages(projectId: string): Promise<StageProgress[]> {
  const res = await api.get<StageProgress[]>(`/v1/projects/${projectId}/stages`)
  return res.data
}

export async function fetchProjectOperationLogs(
  projectId: string,
  limit = 20,
): Promise<OperationLogItem[]> {
  const res = await api.get<OperationLogItem[]>(`/v1/projects/${projectId}/operation-logs`, {
    params: { limit },
  })
  return res.data
}

export async function calculateComplexity(
  payload: ComplexityAssessPayload,
): Promise<ComplexityAssessResult> {
  const res = await api.post<ComplexityAssessResult>('/v1/assess', payload)
  return res.data
}

export async function createSizeEstimate(
  projectId: string,
  payload: SizeEstimatePayload,
): Promise<SizeEstimateResponse> {
  const res = await api.post<SizeEstimateResponse>(
    `/v1/projects/${projectId}/size-estimates`,
    payload,
  )
  return res.data
}

export async function bindSizeEstimate(
  projectId: string,
  estimateId: string | null,
): Promise<{ status: string; project_id: string }> {
  const res = await api.patch(`/v1/projects/${projectId}/size-estimate`, {
    estimate_id: estimateId,
  })
  return res.data
}

export async function fetchApplications(): Promise<PageResponse<ApplicationItem>> {
  const res = await api.get<PageResponse<ApplicationItem>>('/v1/applications?page_size=100')
  return res.data
}
